server/src/main/java/org/elasticsearch/plugins/PluginsLoader.java

```diff
@@ -110,8 +113,12 @@ public static LayerAndLoader ofLoader(ClassLoader loader) {
 
     private final List<PluginDescriptor> moduleDescriptors;
     private final List<PluginDescriptor> pluginDescriptors;
+    private final Set<PluginBundle> allBundles;
     private final Map<String, LoadedPluginLayer> loadedPluginLayers;
 
+    private final Map<ModuleLayer, PluginDescriptor> modularPluginDescriptorByLayer = new HashMap<>();
+    private final Map<ClassLoader, PluginDescriptor> nonModularPluginDescriptorByClassLoader = new HashMap<>();
```

Comment:
These maps can be populated "as we build" LayerAndLoader (as in this PR), but they can also be built afterwards (e.g. in an external class, inside Elasticsearch.java):
```
modularPluginDescriptorByLayer = loadedPluginLayers.entrySet().stream()
   .collect(Collectors.toUnmodifiableMap(
      e -> e.getValue().pluginModuleLayer(),
      e-> e.getValue().pluginBundle().pluginDescriptor())
   );

nonModularPluginDescriptorByClassLoader = loadedPluginLayers.entrySet().stream()
   .collect(Collectors.toUnmodifiableMap(
      e -> e.getValue().pluginClassLoader, 
	  e-> e.getValue().pluginBundle().pluginDescriptor())
   );
```
It's easier to have them here, but it's not necessary (and maybe have them + `resolveClassToPluginName` in a separate utility class is better).
Wdyt?

Reply:
Separated this in a `PluginsResolver` class